# Data Understanding and Preparation

In [188]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Loading the dataset

In [189]:
df = pd.read_csv('../dataset/customer_supermarket.csv', sep='\t', index_col=0)
print(df.tail())

       BasketID      BasketDate  Sale  CustomerID CustomerCountry ProdID  \
541905   581587  09/12/11 12:50   2,1     12680.0          France  22899   
541906   581587  09/12/11 12:50  4,15     12680.0          France  23254   
541907   581587  09/12/11 12:50  4,15     12680.0          France  23255   
541908   581587  09/12/11 12:50  4,95     12680.0          France  22138   
541909   581587  09/12/11 12:50    18     12680.0          France   POST   

                              ProdDescr  Qta  
541905     CHILDREN'S APRON DOLLY GIRL     6  
541906    CHILDRENS CUTLERY DOLLY GIRL     4  
541907  CHILDRENS CUTLERY CIRCUS PARADE    4  
541908    BAKING SET 9 PIECE RETROSPOT     3  
541909                          POSTAGE    1  


## Basic checks

In [190]:
df.drop_duplicates(ignore_index=True, inplace=True)
print(df.tail())

       BasketID      BasketDate  Sale  CustomerID CustomerCountry ProdID  \
466673   581587  09/12/11 12:50   2,1     12680.0          France  22899   
466674   581587  09/12/11 12:50  4,15     12680.0          France  23254   
466675   581587  09/12/11 12:50  4,15     12680.0          France  23255   
466676   581587  09/12/11 12:50  4,95     12680.0          France  22138   
466677   581587  09/12/11 12:50    18     12680.0          France   POST   

                              ProdDescr  Qta  
466673     CHILDREN'S APRON DOLLY GIRL     6  
466674    CHILDRENS CUTLERY DOLLY GIRL     4  
466675  CHILDRENS CUTLERY CIRCUS PARADE    4  
466676    BAKING SET 9 PIECE RETROSPOT     3  
466677                          POSTAGE    1  


In [191]:
df.dtypes

BasketID            object
BasketDate          object
Sale                object
CustomerID         float64
CustomerCountry     object
ProdID              object
ProdDescr           object
Qta                  int64
dtype: object

In [192]:
df['Sale'] = df['Sale'].str.replace(',', '.')
df['Sale'] = pd.to_numeric(df['Sale'])
df = df.astype({'BasketDate': 'datetime64',
                'CustomerID': 'object',
                'CustomerCountry': 'string',
                'ProdDescr': 'string'})

In [193]:
df.dtypes

BasketID                   object
BasketDate         datetime64[ns]
Sale                      float64
CustomerID                 object
CustomerCountry            string
ProdID                     object
ProdDescr                  string
Qta                         int64
dtype: object

In [194]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466678 entries, 0 to 466677
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   BasketID         466678 non-null  object        
 1   BasketDate       466678 non-null  datetime64[ns]
 2   Sale             466678 non-null  float64       
 3   CustomerID       401605 non-null  object        
 4   CustomerCountry  466678 non-null  string        
 5   ProdID           466678 non-null  object        
 6   ProdDescr        465925 non-null  string        
 7   Qta              466678 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(1), object(3), string(2)
memory usage: 28.5+ MB


In [195]:
df.isnull().any()

BasketID           False
BasketDate         False
Sale               False
CustomerID          True
CustomerCountry    False
ProdID             False
ProdDescr           True
Qta                False
dtype: bool

In [201]:
df['ProdDescr'] = (df.groupby('ProdID')['ProdDescr'].transform(lambda x: x[x != ''].iat[0]))

IndexError: index 0 is out of bounds for axis 0 with size 0

'lost' and 'check' in ProdDescr (?)

CustomerID (?) 401605 -> 466678

neg values in Qta (?)

In [160]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466678 entries, 0 to 466677
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   BasketID         466678 non-null  object        
 1   BasketDate       466678 non-null  datetime64[ns]
 2   Sale             466678 non-null  float64       
 3   CustomerID       401605 non-null  object        
 4   CustomerCountry  466678 non-null  string        
 5   ProdID           466678 non-null  object        
 6   ProdDescr        465925 non-null  string        
 7   Qta              466678 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(1), object(3), string(2)
memory usage: 28.5+ MB


In [161]:
df.isnull().any()

BasketID           False
BasketDate         False
Sale               False
CustomerID          True
CustomerCountry    False
ProdID             False
ProdDescr           True
Qta                False
dtype: bool

In [138]:
df.CustomerCountry.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

## Basic statistics

In [132]:
df.describe()

CustomerID            Qta
count  401605.000000  466678.000000
mean    15281.154341      10.806528
std      1714.008869     232.627771
min     12346.000000  -80995.000000
25%     13939.000000       2.000000
50%     15145.000000       4.000000
75%     16784.000000      12.000000
max     18287.000000   80995.000000